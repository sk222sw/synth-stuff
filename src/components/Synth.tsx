import * as React from 'react'
import keys from '../assets/keys.json'
import hoo from '../models/Synth'

import * as R from 'ramda'
import styled from 'styled-components'
import Envelope from '../components/Envelope'
import Filter from '../components/Filter'
import Keyboard from '../components/Keyboard'
import Oscillator from '../components/Oscillator'
import { CenteredRow, Row } from '../components/styles/index'
import { keyExists, keyIsPressed, removeKey } from '../helpers/KeyHandler'
import ComputerKeyboard from '../hocs/ComputerKeyboard'

const StyledSynth = styled.div`
  display: flex;
  flex-direction: column;
  align-self: center;
  width: 616px;
  background: white;
  padding: 8px;
  box-shadow: 0 1px 3px rgba(0,0,0,0.12), 0 1px 2px rgba(0,0,0,0.24);
`

const findByKeyPress = oscillators => R.compose(
  R.filter(R.__, oscillators),
  R.propEq('keyPress'),
)

class Synth extends React.Component<{}, any> {
  constructor(props) {
    super(props)

    this.state = {
      synth: undefined,
      muted: true,
      frequency: 880.00,
      volume: 0.1,
      waveforms: [
        'sine',
        'triangle',
        'sawtooth',
        'square',

      ],
      envelope: {
        a: 10,
        d: 164,
        s: 0.5,
        r: 50,
      },
      pressedKeys: [],
      keys,
      oscillatorConfigs: [
        {
          offset: 0,
          playing: false,
          semi: 0,
          waveform: 'sine',
          peakVolume: 0.5,
        },
        {
          offset: 0,
          playing: false,
          semi: 0,
          waveform: 'sine',
          peakVolume: 0.5,
        },
      ],
      filter: {
        type: 'lowpass',
        frequency: 1000,
      },
    }
  }

  componentWillMount() {
    const ctx = hoo.setup()
    const synth = hoo.createSynth(ctx)()

    window.addEventListener('mouseup', this.stopPlaying)

    const oscillatorConfigs =
      this.state.oscillatorConfigs.map((o, id) => ({ ...o,id }))

    this.setState({ synth, oscillatorConfigs })
  }

  componentWillUnMount() {
    window.removeEventListener('mouseup', this.stopPlaying)
  }

  setOffset = (oscillator) => (offset) => {
    const { oscillatorConfigs } = this.state
    const index = R.findIndex(R.propEq('id', oscillator.id))(oscillatorConfigs)
    const current = oscillatorConfigs[index]

    if (oscillator.playing)
      hoo.setOffset(this.state.synth.oscillators[index], offset)

    this.updateOscillatorConfig(index, current, 'offset', offset)
  }

  setOscillatorVolume = (oscillator) => (volume) => {
    const { oscillatorConfigs } = this.state
    const index = R.findIndex(R.propEq('id', oscillator.id))(oscillatorConfigs)
    const current = oscillatorConfigs[index]

    if (oscillator.playing)
      hoo.setOscillatorVolume(this.state.synth.oscillators[index], volume)

    this.updateOscillatorConfig(index, current, 'peakVolume', volume)
  }

  setWaveform = oscillator => waveform => {
    const { oscillatorConfigs } = this.state
    const index = this.getOscillatorConfigById(oscillator.id)
    const current = oscillatorConfigs[index]

    if (oscillator.playing)
      hoo.setWaveform(this.state.synth.oscillators[index], waveform)

    this.updateOscillatorConfig(index, current, 'waveform', waveform)
  }

  setFilterFrequency = frequency => {
    hoo.setFilterFrequency(this.state.synth.oscillators, frequency)
    this.setState({ filter: { ...this.state.filter, frequency } })
  }

  setSemi = oscillator => value => {
    const semi = Number(value)
    const { oscillatorConfigs } = this.state
    const index = this.getOscillatorConfigById(oscillator.id)
    const current = oscillatorConfigs[index]

    if (oscillator.playing)
      hoo.setSemi(this.state.synth.oscillators[index], semi)

    this.updateOscillatorConfig(index, current, 'semi', semi)
  }

  setFrequency = (frequency) => {
    hoo.setFrequency(this.state.synth, frequency)
    this.setState({ frequency })
  }

  updateOscillatorConfig = (index, oscillator, field, value) => {
    const updated = R.update(
      index,
      { ...oscillator, [field]: value },
      this.state.oscillatorConfigs,
    )

    this.setState({ oscillatorConfigs: updated })
    return updated
  }

  getOscillatorConfigById = id => {
    return R.findIndex(R.propEq('id', id))(this.state.oscillatorConfigs)
  }

  setOscillatorsToPlaying = () => {
    const oscillators = this.state.oscillatorConfigs.map(o => {
      o.playing = true
    })

    this.setState({ oscillators })
  }

  playWithoutPortamento = key => {
    this.state.oscillatorConfigs.forEach(o => {
      hoo.addOscillator(this.state.synth, {
        id: o.id,
        frequency: key.frequency,
        volume: o.peakVolume,
        offset: o.offset,
        waveform: o.waveform,
        keyPress: key.keyPress,
        semi: o.semi,
        filter: this.state.filter,
      })
    })

    this.setOscillatorsToPlaying()

    hoo.play(this.state.synth, this.state.envelope)
  }

  onKeyClick = key => {
    this.handleKeyDown({ key: key.keyPress })
  }

  handleKeyDown = ({ key }) => {
    if (keyExists(key) && !keyIsPressed(this.state.pressedKeys, key)) {

      const keyData = R.find(R.propEq('keyPress', key), (this.state.keys))

      this.playWithoutPortamento(keyData)

      this.setState({
        pressedKeys: [...this.state.pressedKeys, key],
      })
    }
  }

  stopPlaying = () => {
    this.state.pressedKeys.forEach(key => {
      this.handleKeyUp({ key })
    })
  }

  handleKeyUp = ({ key }) => {
    if (keyExists(key)) {
      this.setState({
        pressedKeys: removeKey(this.state.pressedKeys, key),
      })
      const oscillators = findByKeyPress(this.state.synth.oscillators)(key)
      hoo.stopOscillators(this.state.synth, oscillators, this.state.envelope.r)

      this.setState({
        synth: {
          ...this.state.synth,
          oscillators: this.state.synth.oscillators
          .filter(o => o.keyPress !== key),
        },
      })
    }
  }

  changeEnvelope = type => value => {
    const envelope = {
      ...this.state.envelope,
      [type]: value,
    }

    this.setState({ envelope })
  }

  render() {
    return (
      <StyledSynth>
        <ComputerKeyboard
          onKeyDown={this.handleKeyDown}
          onKeyUp={this.handleKeyUp}
        />
        <Row style={{ width: 600 }}>
          {this.state.oscillatorConfigs.map(o =>
            <Oscillator
              key={o.id}
              oscillator={o}
              waveforms={this.state.waveforms}
              setOscillatorVolume={this.setOscillatorVolume}
              setOffset={this.setOffset}
              setSemi={this.setSemi}
              setWaveform={this.setWaveform(o)}
            />,
          )}
        </Row>
        <CenteredRow>
          <Envelope
            envelope={this.state.envelope}
            changeAttack={this.changeEnvelope('a')}
            changeDecay={this.changeEnvelope('d')}
            changeSustain={this.changeEnvelope('s')}
            changeRelease={this.changeEnvelope('r')}
          />
          <Filter
            frequency={this.state.filter.frequency}
            type={this.state.filter.type}
            onFrequencyChange={this.setFilterFrequency}
          />
        </CenteredRow>
        <Keyboard
          keys={this.state.keys}
          currentKeys={this.state.pressedKeys}
          onKeyClick={this.onKeyClick}
        />
      </StyledSynth>
    )
  }
}

export default Synth
