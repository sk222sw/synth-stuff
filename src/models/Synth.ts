import Filter from './Filter'
import Oscillator from './Oscillator'

let ctx

// perhaps remove this and make the user provide a context when creating a new synth?
const setup = (context = new AudioContext()) => {
  return ctx = context
}

// for testing purposes
export const _deleteContext = () =>
  ctx = null

const createSynth = (context = ctx) => ({ oscillators = [] } = {}) => ({
  oscillators,
  context,
})

const addOscillator = (synth, config = {}) => {
  synth.oscillators.push(Oscillator.create(ctx)(config))
  return synth
}

const play = (synth, envelope, filter?) => {
  synth.oscillators.forEach(o => {
    const filterNode = filter
      ? Filter.createFilter(ctx, filter)
      : undefined

    Oscillator.start(ctx, o, ctx.currentTime, envelope, filterNode)
  })
}

const stop = (synth, time) => {
  synth.oscillators.forEach(o => Oscillator.stop(o, time))
  synth.oscillators = []
}

const setFrequency = (synth, frequency) => {
  synth.oscillators.forEach(o => Oscillator.setFrequency(o, frequency))
}

const setOffset = (oscillator, offset) => {
  if (oscillator)
    Oscillator.setOffset(oscillator, offset)
}

const setOscillatorVolume = (oscillator, volume) => {
  if (oscillator)
    Oscillator.setVolume(oscillator, volume)
}

const setSemi = (oscillator, semi) => {
  if (oscillator)
    Oscillator.setSemi(oscillator, semi)
}

const setWaveform = (oscillator, waveform) => {
  if (oscillator)
    Oscillator.setWaveform(oscillator, waveform)
}

const stopOscillators = (synth, oscillators, release) => {
  oscillators.forEach(oscillator => {
    Oscillator.stop(oscillator, ctx.currentTime + release / 1000)
  })
}

const setFilterFrequency = (synth, frequency) => {
  Filter.setFrequency(synth.filter, frequency)
}

export default {
  setup,
  createSynth,
  addOscillator,
  play,
  stop,
  setFrequency,
  setOffset,
  setSemi,
  setWaveform,
  stopOscillators,
  setOscillatorVolume,
  setFilterFrequency,
}
