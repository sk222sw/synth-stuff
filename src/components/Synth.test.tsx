import { shallow } from 'enzyme'
import * as React from 'react'
import 'web-audio-test-api'
import Synth from './Synth'

test('Synth', () => {
  const wrapper = shallow(
    <Synth />,
  )

  expect(wrapper).toMatchSnapshot()
})
