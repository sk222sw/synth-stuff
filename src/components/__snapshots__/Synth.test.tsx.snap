// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Synth 1`] = `
ShallowWrapper {
  "complexSelector": ComplexSelector {
    "buildPredicate": [Function],
    "childrenOfNode": [Function],
    "findWhereUnwrapped": [Function],
  },
  "length": 1,
  "node": <styled.div>
    <ComputerKeyboard
        onKeyDown={[Function]}
        onKeyUp={[Function]}
    />
    <styled.div
        style={
            Object {
                "width": 600,
              }
        }
    >
        <Oscillator
            oscillator={
                Object {
                    "id": 0,
                    "offset": 0,
                    "peakVolume": 0.5,
                    "playing": false,
                    "semi": 0,
                    "waveform": "sine",
                  }
            }
            setOffset={[Function]}
            setOscillatorVolume={[Function]}
            setSemi={[Function]}
            setWaveform={[Function]}
            waveforms={
                Array [
                    "sine",
                    "triangle",
                    "sawtooth",
                    "square",
                  ]
            }
        />
        <Oscillator
            oscillator={
                Object {
                    "id": 1,
                    "offset": 0,
                    "peakVolume": 0.5,
                    "playing": false,
                    "semi": 0,
                    "waveform": "sine",
                  }
            }
            setOffset={[Function]}
            setOscillatorVolume={[Function]}
            setSemi={[Function]}
            setWaveform={[Function]}
            waveforms={
                Array [
                    "sine",
                    "triangle",
                    "sawtooth",
                    "square",
                  ]
            }
        />
    </styled.div>
    <Styled(styled.div)>
        <Envelope
            changeAttack={[Function]}
            changeDecay={[Function]}
            changeRelease={[Function]}
            changeSustain={[Function]}
            envelope={
                Object {
                    "a": 10,
                    "d": 164,
                    "r": 50,
                    "s": 0.5,
                  }
            }
        />
        <Filter
            frequency={1000}
            onFrequencyChange={[Function]}
            type="lowpass"
        />
    </Styled(styled.div)>
    <Keyboard
        currentKeys={Array []}
        keys={undefined}
        onKeyClick={[Function]}
    />
</styled.div>,
  "nodes": Array [
    <styled.div>
      <ComputerKeyboard
            onKeyDown={[Function]}
            onKeyUp={[Function]}
      />
      <styled.div
            style={
                  Object {
                        "width": 600,
                      }
            }
      >
            <Oscillator
                  oscillator={
                        Object {
                              "id": 0,
                              "offset": 0,
                              "peakVolume": 0.5,
                              "playing": false,
                              "semi": 0,
                              "waveform": "sine",
                            }
                  }
                  setOffset={[Function]}
                  setOscillatorVolume={[Function]}
                  setSemi={[Function]}
                  setWaveform={[Function]}
                  waveforms={
                        Array [
                              "sine",
                              "triangle",
                              "sawtooth",
                              "square",
                            ]
                  }
            />
            <Oscillator
                  oscillator={
                        Object {
                              "id": 1,
                              "offset": 0,
                              "peakVolume": 0.5,
                              "playing": false,
                              "semi": 0,
                              "waveform": "sine",
                            }
                  }
                  setOffset={[Function]}
                  setOscillatorVolume={[Function]}
                  setSemi={[Function]}
                  setWaveform={[Function]}
                  waveforms={
                        Array [
                              "sine",
                              "triangle",
                              "sawtooth",
                              "square",
                            ]
                  }
            />
      </styled.div>
      <Styled(styled.div)>
            <Envelope
                  changeAttack={[Function]}
                  changeDecay={[Function]}
                  changeRelease={[Function]}
                  changeSustain={[Function]}
                  envelope={
                        Object {
                              "a": 10,
                              "d": 164,
                              "r": 50,
                              "s": 0.5,
                            }
                  }
            />
            <Filter
                  frequency={1000}
                  onFrequencyChange={[Function]}
                  type="lowpass"
            />
      </Styled(styled.div)>
      <Keyboard
            currentKeys={Array []}
            keys={undefined}
            onKeyClick={[Function]}
      />
</styled.div>,
  ],
  "options": Object {},
  "renderer": ReactShallowRenderer {
    "_instance": ShallowComponentWrapper {
      "_calledComponentWillUnmount": false,
      "_compositeType": 0,
      "_context": Object {},
      "_currentElement": <Synth />,
      "_debugID": 1,
      "_hostContainerInfo": null,
      "_hostParent": null,
      "_instance": Synth {
        "_reactInternalInstance": [Circular],
        "changeEnvelope": [Function],
        "context": Object {},
        "getOscillatorConfigById": [Function],
        "handleKeyDown": [Function],
        "handleKeyUp": [Function],
        "onKeyClick": [Function],
        "playWithoutPortamento": [Function],
        "props": Object {},
        "refs": Object {},
        "setFilterFrequency": [Function],
        "setFrequency": [Function],
        "setOffset": [Function],
        "setOscillatorVolume": [Function],
        "setOscillatorsToPlaying": [Function],
        "setSemi": [Function],
        "setWaveform": [Function],
        "state": Object {
          "envelope": Object {
            "a": 10,
            "d": 164,
            "r": 50,
            "s": 0.5,
          },
          "filter": Object {
            "frequency": 1000,
            "type": "lowpass",
          },
          "frequency": 880,
          "keys": undefined,
          "muted": true,
          "oscillatorConfigs": Array [
            Object {
              "id": 0,
              "offset": 0,
              "peakVolume": 0.5,
              "playing": false,
              "semi": 0,
              "waveform": "sine",
            },
            Object {
              "id": 1,
              "offset": 0,
              "peakVolume": 0.5,
              "playing": false,
              "semi": 0,
              "waveform": "sine",
            },
          ],
          "pressedKeys": Array [],
          "synth": Object {
            "context": Object {
              "inputs": Array [],
              "name": "AudioDestinationNode",
            },
            "oscillators": Array [],
          },
          "volume": 0.1,
          "waveforms": Array [
            "sine",
            "triangle",
            "sawtooth",
            "square",
          ],
        },
        "stopPlaying": [Function],
        "updateOscillatorConfig": [Function],
        "updater": Object {
          "enqueueCallback": [Function],
          "enqueueCallbackInternal": [Function],
          "enqueueElementInternal": [Function],
          "enqueueForceUpdate": [Function],
          "enqueueReplaceState": [Function],
          "enqueueSetState": [Function],
          "isMounted": [Function],
          "validateCallback": [Function],
        },
      },
      "_mountOrder": 1,
      "_pendingCallbacks": null,
      "_pendingElement": null,
      "_pendingForceUpdate": false,
      "_pendingReplaceState": false,
      "_pendingStateQueue": null,
      "_renderedComponent": NoopInternalComponent {
        "_currentElement": <styled.div>
          <ComputerKeyboard
                    onKeyDown={[Function]}
                    onKeyUp={[Function]}
          />
          <styled.div
                    style={
                              Object {
                                        "width": 600,
                                      }
                    }
          >
                    <Oscillator
                              oscillator={
                                        Object {
                                                  "id": 0,
                                                  "offset": 0,
                                                  "peakVolume": 0.5,
                                                  "playing": false,
                                                  "semi": 0,
                                                  "waveform": "sine",
                                                }
                              }
                              setOffset={[Function]}
                              setOscillatorVolume={[Function]}
                              setSemi={[Function]}
                              setWaveform={[Function]}
                              waveforms={
                                        Array [
                                                  "sine",
                                                  "triangle",
                                                  "sawtooth",
                                                  "square",
                                                ]
                              }
                    />
                    <Oscillator
                              oscillator={
                                        Object {
                                                  "id": 1,
                                                  "offset": 0,
                                                  "peakVolume": 0.5,
                                                  "playing": false,
                                                  "semi": 0,
                                                  "waveform": "sine",
                                                }
                              }
                              setOffset={[Function]}
                              setOscillatorVolume={[Function]}
                              setSemi={[Function]}
                              setWaveform={[Function]}
                              waveforms={
                                        Array [
                                                  "sine",
                                                  "triangle",
                                                  "sawtooth",
                                                  "square",
                                                ]
                              }
                    />
          </styled.div>
          <Styled(styled.div)>
                    <Envelope
                              changeAttack={[Function]}
                              changeDecay={[Function]}
                              changeRelease={[Function]}
                              changeSustain={[Function]}
                              envelope={
                                        Object {
                                                  "a": 10,
                                                  "d": 164,
                                                  "r": 50,
                                                  "s": 0.5,
                                                }
                              }
                    />
                    <Filter
                              frequency={1000}
                              onFrequencyChange={[Function]}
                              type="lowpass"
                    />
          </Styled(styled.div)>
          <Keyboard
                    currentKeys={Array []}
                    keys={undefined}
                    onKeyClick={[Function]}
          />
</styled.div>,
        "_debugID": 2,
        "_renderedOutput": <styled.div>
          <ComputerKeyboard
                    onKeyDown={[Function]}
                    onKeyUp={[Function]}
          />
          <styled.div
                    style={
                              Object {
                                        "width": 600,
                                      }
                    }
          >
                    <Oscillator
                              oscillator={
                                        Object {
                                                  "id": 0,
                                                  "offset": 0,
                                                  "peakVolume": 0.5,
                                                  "playing": false,
                                                  "semi": 0,
                                                  "waveform": "sine",
                                                }
                              }
                              setOffset={[Function]}
                              setOscillatorVolume={[Function]}
                              setSemi={[Function]}
                              setWaveform={[Function]}
                              waveforms={
                                        Array [
                                                  "sine",
                                                  "triangle",
                                                  "sawtooth",
                                                  "square",
                                                ]
                              }
                    />
                    <Oscillator
                              oscillator={
                                        Object {
                                                  "id": 1,
                                                  "offset": 0,
                                                  "peakVolume": 0.5,
                                                  "playing": false,
                                                  "semi": 0,
                                                  "waveform": "sine",
                                                }
                              }
                              setOffset={[Function]}
                              setOscillatorVolume={[Function]}
                              setSemi={[Function]}
                              setWaveform={[Function]}
                              waveforms={
                                        Array [
                                                  "sine",
                                                  "triangle",
                                                  "sawtooth",
                                                  "square",
                                                ]
                              }
                    />
          </styled.div>
          <Styled(styled.div)>
                    <Envelope
                              changeAttack={[Function]}
                              changeDecay={[Function]}
                              changeRelease={[Function]}
                              changeSustain={[Function]}
                              envelope={
                                        Object {
                                                  "a": 10,
                                                  "d": 164,
                                                  "r": 50,
                                                  "s": 0.5,
                                                }
                              }
                    />
                    <Filter
                              frequency={1000}
                              onFrequencyChange={[Function]}
                              type="lowpass"
                    />
          </Styled(styled.div)>
          <Keyboard
                    currentKeys={Array []}
                    keys={undefined}
                    onKeyClick={[Function]}
          />
</styled.div>,
      },
      "_renderedNodeType": 1,
      "_rootNodeID": 0,
      "_topLevelWrapper": null,
      "_updateBatchNumber": null,
      "_warnedAboutRefsInRender": false,
    },
    "getRenderOutput": [Function],
    "render": [Function],
  },
  "root": [Circular],
  "unrendered": <Synth />,
}
`;
